src/index.ts

```diff
@@ -1,102 +1,225 @@
-import * as core from '@actions/core';
-import * as exec from '@actions/exec';
-import * as io from '@actions/io';
+import * as core from "@actions/core";
+import * as exec from "@actions/exec";
+import * as io from "@actions/io";
 import * as fs from "fs";
 import * as path from "path";
 
-export async function run(): Promise<void> {
-    const imageInput = core.getInput('image', { required: true });
-    const tag = core.getInput('tag') || 'latest';
-    const registry = core.getInput('registry', { required: true });
-    const username = core.getInput('username', { required: true });
-    const password = core.getInput('password', { required: true });
-    const tlsVerify = core.getInput('tls-verify');
-    const digestFileInput = core.getInput('digestfile');
+interface ExecResult {
+    exitCode: number;
+    stdout: string;
+    stderr: string;
+}
 
-    // get podman cli
-    const podman = await io.which('podman', true);
+// base URL that gets appended if image is pulled from the Docker storage
+const dockerBaseUrl = "docker.io/library";
 
-    const imageToPush = `${imageInput}:${tag}`;
-    let pushMsg = `Pushing ${imageToPush} to ${registry}`;
-    if (username) {
-        pushMsg += ` as ${username}`;
+async function run(): Promise<void> {
+    const imageInput = core.getInput("image", { required: true });
+    const tag = core.getInput("tag") || "latest";
+    const registry = core.getInput("registry", { required: true });
+    const username = core.getInput("username", { required: true });
+    const password = core.getInput("password", { required: true });
+    const tlsVerify = core.getInput("tls-verify");
+    const digestFileInput = core.getInput("digestfile");
+
+    // get Podman cli path
+    const podmanPath = await io.which("podman", true);
+
+    let imageToPush = `${imageInput}:${tag}`;
+
+    // check if image exist in Podman storage
+    const isPresentInPodman: boolean = await checkImageInPodman(
+        imageToPush,
+        podmanPath,
+    );
+
+    // check if image exist in Docker storage and if exist pull the image to Podman
+    const isPresentInDocker: boolean = await pullImageFromDocker(
+        imageToPush,
+        podmanPath,
+    );
+
+    // failing if image is not found in Docker as well as Podman
+    if (!isPresentInDocker && !isPresentInPodman) {
+        core.setFailed(`Image ${imageToPush} not found in local storage's of Podman or Docker`);
```

Comment:
`throw new Error` instead of `setFailed` and `return`

`Image "${imageToPush} not found in Podman local storage, or Docker local storage.`